Change symbolic::Variable to have shared_ptr<string> name_

soonho-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5983)
<!-- Reviewable:end -->

soonho-tri commented:
+@ericcousineau-tri , feature-review please

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-:-Kj3VF-b-JTDHzB9KSEl:bu743v2)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/symbolic_variable.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj3zwH1yuk3jf20Du9q:-Kj3zwH1yuk3jf20Du9r:bcxf43k) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/symbolic_variable.cc#L30)):*
> ```cpp
> 
> Variable::Variable(const string& name)
>     : id_{get_next_id()}, name_{make_shared<string>(name)} {
> ```

Is there a reason why we couldn't do `make_shared<string>(std::move(name))`, and keep the value semantics from before?

---

*[drake/common/symbolic_variable.h, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj3zCa62O3N7UEHhknj:-Kj3zCa74ZsXSiH7VnS_:b-e3ubmv) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/symbolic_variable.h#L58)):*
Is the member initialization necessary for `shared_ptr`? Per the [style guide](http://drake.mit.edu/code_style_guide.html#additional-rules):

> Class objects are responsible for their own construction so they do not need to be member-initialized.

---

*[drake/common/test/symbolic_expression_test.cc, line 1765 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4-4PVd6fWqlHCOcqs:-Kj4-4PVd6fWqlHCOcqt:b-22v0m4) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/test/symbolic_expression_test.cc#L1765)):*
> ```cpp
> TEST_F(SymbolicExpressionTest, MemcpyKeepsIntact) {
>   for (const Variable& var : {var_x_, var_y_, var_z_}) {
>     Variable* new_var{static_cast<Variable*>(malloc(sizeof(Variable)))};
> ```

As we discussed before, would it be possible to separate the memory allocation from the casting, to make it easier to see that we won't be calling the destructor on the `memcpy`'d variable?

---

*[drake/common/test/symbolic_expression_test.cc, line 1771 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4-Hy4ZAnHHzVuVcIj:-Kj4-Hy5df0RQ07H5SqJ:b7ucvux) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/test/symbolic_expression_test.cc#L1771)):*
> ```cpp
>   }
>   for (const Expression& expr : collection_) {
>     Expression* new_expr{static_cast<Expression*>(malloc(sizeof(Expression)))};
> ```

Could we split this check into a separate function, such as `ExpectTriviallyMovable`? And is it possible to use `operator==` rather than a specific predicate? (just to minimize specializations and all that)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/symbolic_variable.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj3zwH1yuk3jf20Du9q:-Kj41q8UUPC3zQpxn6N0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/symbolic_variable.cc#L30)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Is there a reason why we couldn't do `make_shared<string>(std::move(name))`, and keep the value semantics from before?
</blockquote></details>

Done.

---

*[drake/common/symbolic_variable.h, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj3zCa62O3N7UEHhknj:-Kj41s-JnXJx1o4jkK1r:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/symbolic_variable.h#L58)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Is the member initialization necessary for `shared_ptr`? Per the [style guide](http://drake.mit.edu/code_style_guide.html#additional-rules):

> Class objects are responsible for their own construction so they do not need to be member-initialized.
</blockquote></details>

Done.

---

*[drake/common/test/symbolic_expression_test.cc, line 1765 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4-4PVd6fWqlHCOcqs:-Kj41wiRlpR02zaxvmBZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/test/symbolic_expression_test.cc#L1765)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

As we discussed before, would it be possible to separate the memory allocation from the casting, to make it easier to see that we won't be calling the destructor on the `memcpy`'d variable?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/test/symbolic_expression_test.cc, line 1771 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4-Hy4ZAnHHzVuVcIj:-Kj43NqDhFBTk2BxptTs:b-lmco6b) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/test/symbolic_expression_test.cc#L1771)):*
> And is it possible to use operator== rather than a specific predicate? (just to minimize specializations and all that)

Since `e1 == e1` produces a symbolic Formula instead of `bool`, we need to use `e1.EqualTo(e2)` if we don't want to use `ExprEqual` predicate.

> Could we split this check into a separate function, such as ExpectTriviallyMovable?

I'd separate it if I can re-use the function with `Variable`, `Expression`, and `Formula`. But it seems that it's not trivial (for example, `Variable` has `equal_to` method for structural equality checking while Expression/Formula have `EqualTo` method).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/test/symbolic_expression_test.cc, line 1771 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4-Hy4ZAnHHzVuVcIj:-Kj4QmU_M6QSlekh5IDz:b-f74yvd) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/test/symbolic_expression_test.cc#L1771)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> And is it possible to use operator== rather than a specific predicate? (just to minimize specializations and all that)

Since `e1 == e1` produces a symbolic Formula instead of `bool`, we need to use `e1.EqualTo(e2)` if we don't want to use `ExprEqual` predicate.

> Could we split this check into a separate function, such as ExpectTriviallyMovable?

I'd separate it if I can re-use the function with `Variable`, `Expression`, and `Formula`. But it seems that it's not trivial (for example, `Variable` has `equal_to` method for structural equality checking while Expression/Formula have `EqualTo` method).
</blockquote></details>

Thanks for the code snippet! I actually realize that we can use `std::equal_to` for all three cases! PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 4 of 4 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/common/symbolic_variable.h, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4pNe6vhre459OXWL0:-Kj4pNe6vhre459OXWL1:b4s18yg) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/symbolic_variable.h#L58)):*
> ```objc
>   static Id get_next_id();
>   Id id_{};                            // Unique identifier.
>   std::shared_ptr<std::string> name_;  // Name of variable.
> ```

Can you reference the issue that motivated the usage of `shared_ptr<string>`versus a plain `string`?

---

*[drake/common/test/is_memcpy_movable.h, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4pBLIPgHHjr3vp7am:-Kj4pBLJKoxvL3I-lx62:bldc6zl) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable.h#L11)):*
> ```objc
> /// it.
> ///
> /// @note @p invariant_pred is a binary function object.
> ```

Can you also add a reference to the issue or relevant discussion that motivated this method?

Do you also think it would be appropriate to say that this should _always_ be used to test a scalar used within an `Eigen::EigenBase<...>` object?

---

*[drake/common/test/is_memcpy_movable.h, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4pvFOfNWKJD3XzP--:-Kj4pvFPZHz8kS1fgQmJ:b-cwoejn) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable.h#L12)):*
> ```objc
> ///
> /// @note @p invariant_pred is a binary function object.
> template <typename T, typename Invariant_Pred = std::equal_to<T>>
> ```

Could you change `Invariant_Pred` to `InvariantPred`? (sorry if I miscommunicated that format in the DM)

---

*[drake/common/test/is_memcpy_movable.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4ph_dcQSTqwwwYWNF:-Kj4ph_dcQSTqwwwYWNG:b-ez0g1q) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable.h#L16)):*
> ```objc
>                      const Invariant_Pred& invariant_pred = Invariant_Pred()) {
>   // 1. Create ptr_to_new_value1 via placement-new.
>   void* const raw1{malloc(sizeof(T))};
> ```

Could you rename `raw1` to `raw_original`, and `raw2` to `raw_moved` or `raw_realloc`, and change the relevant suffixes?

---

*[drake/common/test/is_memcpy_movable_test.cc, line 37 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4qP7yP8eaDy21Hm_e:-Kj4qP7yP8eaDy21Hm_f:b-xsri3) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable_test.cc#L37)):*
> ```cpp
> };
> 
> GTEST_TEST(IsMemcpyMovableTest, NotMemcpyMovableTest) {
> ```

Could you introduce a positive test as well? I know that you have it in downstream examples, but it would be nice to have it here too.

---

*[drake/common/test/symbolic_expression_test.cc, line 1771 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4-Hy4ZAnHHzVuVcIj:-Kj4q56KzGHutA-rZ4tA:b-kbnish) ([raw file](https://github.com/robotlocomotion/drake/blob/496309e7e76c4c87e6dd55a1a5369976a754c60d/drake/common/test/symbolic_expression_test.cc#L1771)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Thanks for the code snippet! I actually realize that we can use `std::equal_to` for all three cases! PTAL.
</blockquote></details>

BTW Welcome!

---

*[drake/common/test/symbolic_expression_test.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4p-cAZ8HbWOGwF0_x:-Kj4p-cC1As-Ldl_RDpi:b-ksyqrg) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/symbolic_expression_test.cc#L51)):*
> ```cpp
> using test::ExprNotEqual;
> using test::ExprNotLess;
> using test::VarEqual;
> ```

Is `VarEqual` still necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
*[drake/common/test/is_memcpy_movable_test.cc, line 37 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4qP7yP8eaDy21Hm_e:-Kj4quNkK-6av4oV1GaR:b-yaid4r) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable_test.cc#L37)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Could you introduce a positive test as well? I know that you have it in downstream examples, but it would be nice to have it here too.
</blockquote></details>

Perhaps something as simple as an `int`, or some trivial-constructable object, such as `struct ... { int x_; const char* name_; };`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
*[drake/common/test/is_memcpy_movable.h, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4pBLIPgHHjr3vp7am:-Kj8XLjwJHIDuFIBp9fb:b-sbtien) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable.h#L11)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Can you also add a reference to the issue or relevant discussion that motivated this method?

Do you also think it would be appropriate to say that this should _always_ be used to test a scalar used within an `Eigen::EigenBase<...>` object?
</blockquote></details>

Could you also add to the Doxygen `@brief` portion that it uses the `invariant_pred` predicate to check for invariance? (just to have it inline as well)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 5 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/common/symbolic_variable.h, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4pNe6vhre459OXWL0:-Kj8T6_AfN9U1qv8tTu2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/symbolic_variable.h#L58)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Can you reference the issue that motivated the usage of `shared_ptr<string>`versus a plain `string`?
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable.h, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4pBLIPgHHjr3vp7am:-Kj8UNkTd3K8KXVayJu2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable.h#L11)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Can you also add a reference to the issue or relevant discussion that motivated this method?

Do you also think it would be appropriate to say that this should _always_ be used to test a scalar used within an `Eigen::EigenBase<...>` object?
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable.h, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4pvFOfNWKJD3XzP--:-Kj8Ud-bQaSnz8HiGTmZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable.h#L12)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Could you change `Invariant_Pred` to `InvariantPred`? (sorry if I miscommunicated that format in the DM)
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4ph_dcQSTqwwwYWNF:-Kj8V4U1wzbQ-tZbXEIn:bs7wumb) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable.h#L16)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Could you rename `raw1` to `raw_original`, and `raw2` to `raw_moved` or `raw_realloc`, and change the relevant suffixes?
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 37 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4qP7yP8eaDy21Hm_e:-Kj8ZQQVRwCtP7OZ9ael:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/is_memcpy_movable_test.cc#L37)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Perhaps something as simple as an `int`, or some trivial-constructable object, such as `struct ... { int x_; const char* name_; };`.
</blockquote></details>

Done.

---

*[drake/common/test/symbolic_expression_test.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj4p-cAZ8HbWOGwF0_x:-Kj8VBm0XrVNLy0RAeGj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/942cea5388d4b1861f057376c9c557ba0e69e44e/drake/common/test/symbolic_expression_test.cc#L51)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Is `VarEqual` still necessary?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
*[drake/common/test/is_memcpy_movable_test.cc, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj94GW7VNFU7bnGzauK:-Kj94GW9g91TGZfQ4BgS:bc3rxev) ([raw file](https://github.com/robotlocomotion/drake/blob/94e08d4fbbae6d1f907a3899917905fdce93d2b4/drake/common/test/is_memcpy_movable_test.cc#L20)):*
> ```cpp
>   MemcpyMovable(int id, string name) : id_{id}, name_{new string{move(name)}} {}
>   MemcpyMovable(const MemcpyMovable& m) : MemcpyMovable{m.id_, *m.name_} {}
>   MemcpyMovable(MemcpyMovable&& m) = default;
> ```

Per our DM convo on Slack, would you be able to imply ownership either by making `name_` a `shared_ptr` or use a non-default move constructor to invalidate `name_`(set it to NULL)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
+@jwnimmer-tri , platform-review please. 

---

Review status: 4 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj94GW7VNFU7bnGzauK:-Kj9BP6T1_mdfzVme-dg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/94e08d4fbbae6d1f907a3899917905fdce93d2b4/drake/common/test/is_memcpy_movable_test.cc#L20)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Per our DM convo on Slack, would you be able to imply ownership either by making `name_` a `shared_ptr` or use a non-default move constructor to invalidate `name_`(set it to NULL)?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-:-Kj8cYWXalAb4Byrq-yt:bvuj2s1)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 9 files at r2, 3 of 4 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/common/symbolic_variable.h, line 32 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9FIyq-HCnXt04O4Bi:-Kj9FIyr-RE45wJzHVIH:btjcvae) ([raw file](https://github.com/robotlocomotion/drake/blob/a7e36f405001b5f45e179ab3b716688b87d66796/drake/common/symbolic_variable.h#L32)):*
> ```objc
>    *  construct a symbolic expression.
>    */
>   Variable() : id_{0}, name_{std::string()} {}
> ```

With this change, `Variable().get_name()` and `.to_string()` and `operator<<` will segfault.  At worst, it should be a DRAKE_ASSERT; but really, it seems like keeping the empty name was good advice?  And if so, I guess some unit tests are missing.

---

*[drake/common/test/is_memcpy_movable.h, line 15 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9G3GhZVpMrntoSNOy:-Kj9G3GhZVpMrntoSNOz:buymnly) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable.h#L15)):*
> ```objc
> /// using `memcpy` without calling a move constructor. As a result, if a
> /// `Scalar` type of Eigen matrix/array is not IsMemcpyMovable, we can have
> /// memory leaks when `conservativeResize` method is called. This should
> ```

BTW its worse than just memory leaks -- its undefined behavior likely to lead to traps.

---

*[drake/common/test/is_memcpy_movable.h, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9GZ6wfrr4uACDOByz:-Kj9GZ6xyH8aTkcX-OxY:b-gmwrku) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable.h#L23)):*
> ```objc
>                      const InvariantPred& invariant_pred = InvariantPred()) {
>   // 1. Create ptr_to_original via placement-new.
>   void* const raw_original{malloc(sizeof(T))};
> ```

Doing this manually can lead to sizing and alignment problems.  Perhaps use something like https://github.com/RobotLocomotion/drake/blob/master/drake/common/never_destroyed.h#L44 as in:
```c++
auto original_storage = make_unique<aligned_storage<...>>();
```

Ditto for the malloc below.

This also gives a nice RAII so we don't need the free-and-end-of-function and can do clear things like `raw_original.reset()` below to free the original.

---

*[drake/common/test/is_memcpy_movable.h, line 32 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9H4vfbc5tn3PZhqBF:-Kj9H4vfbc5tn3PZhqBG:b-zeulwi) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable.h#L32)):*
> ```objc
> 
>   // 3. Destroy ptr_to_original and free raw_original.
>   ptr_to_original->T::~T();
> ```

Actually this is not right.  We should free the memory but not call T's dtor, because we want the _moved_ object to be the live one, not the original object, so we want to ensure that the dtor on the moved object still works, and we should only be calling one of the dtors.  The ptr_to_moved is what we should be calling placement-delete on.  (Also, dtor-on-moved matches what Eigen does internally.)

---

*[drake/common/test/is_memcpy_movable_test.cc, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9JsMT8JrzwgBtACF1:-Kj9JsMT8JrzwgBtACF2:bikpcpc) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L23)):*
> ```cpp
>   MemcpyMovable(int id, string name)
>       : id_{id}, name_{make_shared<string>(move(name))} {}
>   MemcpyMovable(const MemcpyMovable& m) : MemcpyMovable{m.id_, *m.name_} {}
> ```

If its important to induce a `make_shared` here with a copy of the string, then comment that.  If not, then why not use `= default`?

---

*[drake/common/test/is_memcpy_movable_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9JkEzh82jFyl2-XSA:-Kj9JkF-R3K1IjeRc8WQ:b8ulppm) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L37)):*
> ```cpp
> 
>  private:
>   const int id_{};
> ```

I am not sure that having `id` as a member is buying us much?

---

*[drake/common/test/is_memcpy_movable_test.cc, line 38 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9KCSmJxU65kp4xW0M:-Kj9KCSnTK71WH_ksZqt:b-8t4aoi) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L38)):*
> ```cpp
>  private:
>   const int id_{};
>   const shared_ptr<string> name_;
> ```

I wonder if this test case would be easier to understand if this was a `unique_ptr`?  The memcpy-moveable should equally apply to any class that holds a simple pointer; the shared control block semantics don't really enter into it.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 49 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9Kb51KuxuJKyDDeG3:-Kj9Kb51KuxuJKyDDeG4:b5mx1dm) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L49)):*
> ```cpp
> 
>   void operator=(const NotMemcpyMovable&) {}
>   void operator=(NotMemcpyMovable&&) {}
> ```

Why do these two methods exist?  The prior class didn't appear to need them.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 76 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9IBhys_0BnQcK5cwp:-Kj9IBhys_0BnQcK5cwq:bvg9rp0) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L76)):*
> ```cpp
> 
> GTEST_TEST(IsMemcpyMovableTest, MemcpyMovableTestStringPtr) {
>   const string* const v{new string{"3.14"}};
> ```

FYI It seems like keeping `v` in a `unique_ptr` would be just as good of a test, and then we don't need the manual `delete`.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 84 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9IROXmXkWYEczvmHo:-Kj9IROXmXkWYEczvmHp:bqs366a) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L84)):*
> ```cpp
>   const MemcpyMovable v{5, "MemcpyMovable"};
>   EXPECT_TRUE(IsMemcpyMovable(v, MemcpyMovable::EqualTo()));
>   EXPECT_TRUE(IsMemcpyMovable(move(v), MemcpyMovable::EqualTo()));
> ```

BTW This is a bit weird -- the function doesn't deal with moved arguments at all, does it?

---

*[drake/common/test/symbolic_variable_test.cc, line 167 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9EMSFlBGF64hKOFmW:-Kj9EMSFlBGF64hKOFmX:b-yciaj4) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/symbolic_variable_test.cc#L167)):*
> ```cpp
>   for (const Variable& var : {x_, y_, z_, w_}) {
>     EXPECT_TRUE(IsMemcpyMovable(var));
>   }
> ```

Add test cases for Variable names of differing lengths?  (At least, strings that are longer than 16ish characters.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 5 of 9 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/common/symbolic_variable.h, line 32 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9FIyq-HCnXt04O4Bi:-Kj9M-NgHhGnAs4z2cpa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a7e36f405001b5f45e179ab3b716688b87d66796/drake/common/symbolic_variable.h#L32)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

With this change, `Variable().get_name()` and `.to_string()` and `operator<<` will segfault.  At worst, it should be a DRAKE_ASSERT; but really, it seems like keeping the empty name was good advice?  And if so, I guess some unit tests are missing.
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable.h, line 15 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9G3GhZVpMrntoSNOy:-Kj9MZy7cMoVcLzMZ0QY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable.h#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW its worse than just memory leaks -- its undefined behavior likely to lead to traps.
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable.h, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9GZ6wfrr4uACDOByz:-Kj9gMbAOKNDSAGziWIT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable.h#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Doing this manually can lead to sizing and alignment problems.  Perhaps use something like https://github.com/RobotLocomotion/drake/blob/master/drake/common/never_destroyed.h#L44 as in:
```c++
auto original_storage = make_unique<aligned_storage<...>>();
```

Ditto for the malloc below.

This also gives a nice RAII so we don't need the free-and-end-of-function and can do clear things like `raw_original.reset()` below to free the original.
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable.h, line 32 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9H4vfbc5tn3PZhqBF:-Kj9gOGyk56nQ9X4v9tZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable.h#L32)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Actually this is not right.  We should free the memory but not call T's dtor, because we want the _moved_ object to be the live one, not the original object, so we want to ensure that the dtor on the moved object still works, and we should only be calling one of the dtors.  The ptr_to_moved is what we should be calling placement-delete on.  (Also, dtor-on-moved matches what Eigen does internally.)
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj94GW7VNFU7bnGzauK:-Kj9iVuvgY-mBvQZ_gJR:bnqh0hk) ([raw file](https://github.com/robotlocomotion/drake/blob/94e08d4fbbae6d1f907a3899917905fdce93d2b4/drake/common/test/is_memcpy_movable_test.cc#L20)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Now it's using `unique_ptr`.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9JsMT8JrzwgBtACF1:-Kj9gjuga70DuAj8iyER:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If its important to induce a `make_shared` here with a copy of the string, then comment that.  If not, then why not use `= default`?
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9JkEzh82jFyl2-XSA:-Kj9gmvo-lQIyc4ojqnw:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L37)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am not sure that having `id` as a member is buying us much?
</blockquote></details>

Removed.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 38 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9KCSmJxU65kp4xW0M:-Kj9goVjNAbH8eLI7V5z:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L38)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I wonder if this test case would be easier to understand if this was a `unique_ptr`?  The memcpy-moveable should equally apply to any class that holds a simple pointer; the shared control block semantics don't really enter into it.
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 49 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9Kb51KuxuJKyDDeG3:-Kj9UXBJCpCKQ2JRq8-x:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L49)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Why do these two methods exist?  The prior class didn't appear to need them.
</blockquote></details>

Removed.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 76 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9IBhys_0BnQcK5cwp:-Kj9h7ela_6hSreQpTqb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L76)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI It seems like keeping `v` in a `unique_ptr` would be just as good of a test, and then we don't need the manual `delete`.
</blockquote></details>

Done.

---

*[drake/common/test/is_memcpy_movable_test.cc, line 84 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9IROXmXkWYEczvmHo:-Kj9hFMJvfigc1kMe0hW:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/is_memcpy_movable_test.cc#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This is a bit weird -- the function doesn't deal with moved arguments at all, does it?
</blockquote></details>

Removed.

---

*[drake/common/test/symbolic_variable_test.cc, line 167 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-Kj9EMSFlBGF64hKOFmW:-Kj9jgsP0lggDTxuRqSi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22fdb8b6b16003cab5ce154a3c45985979816040/drake/common/test/symbolic_variable_test.cc#L167)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Add test cases for Variable names of differing lengths?  (At least, strings that are longer than 16ish characters.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-:-Kj9qJBleBFtXNcpzvxu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5983#-:-Kj9v6jhjD3GzBhgbygs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Making mention of #5974.
